### DISCUSSION TITLE
multi_ray_tracing speed on different Python Versions

### DISCUSSION THREAD

**Discussion by danbodily** [pyvista/pyvista#4369] [Q&A]
I've been doing some tests using multi_ray_tracing with a very simple script.

```py
import pyvista as pv
sphere = pv.Sphere()
import numpy as np

starts  = np.zeros([3,10000]).T
ends = np.random.random([3,10000]).T

import time
t = time.time()
points, rays, cells = sphere.multi_ray_trace(starts, ends, first_point=True)..
print("TIME: ", time.time() - t, flush = True)

t = time.time()
points, rays, cells = sphere.multi_ray_trace(starts, ends, first_point=True)..
print("SECOND TIME: ", time.time() - t, flush = True)
```

I'm finding this chunk of code executes almost 100x faster in Python 3.8.16 (after the first time when the tree is set up) than in Python 3.8.10  (obtaining these python versions through Miniconda).  I don't see these same speedups in 3.9 or 3.10 python versions (same as 3.8.10).   Any ideas why 3.8.16 is so much faster?  Both pip freeze in 3.8.10 and 3.8.16 return identical packages:

```
certifi==2022.12.7
charset-normalizer==3.1.0
contourpy==1.0.7
cycler==0.11.0
Cython==0.29.34
fonttools==4.39.3
idna==3.4
imageio==2.28.1
importlib-resources==5.12.0
kiwisolver==1.4.4
matplotlib==3.7.1
numpy==1.24.3
packaging==23.1
Pillow==9.5.0
platformdirs==3.5.0
pooch==1.7.0
pyembree==0.2.11
pyparsing==3.0.9
python-dateutil==2.8.2
pyvista==0.38.5
requests==2.29.0
Rtree==1.0.1
scooby==0.7.1
six==1.16.0
trimesh==3.21.5
urllib3==1.26.15
vtk==9.2.6
zipp==3.15.0
```

**Comment by banesullivan** [comment]
I have no idea why there would be such a dramatic speedup. I recommend using a an external tool to track execution times in a waterfall graph to see where the difference is happening. [This blog](https://roman.pt/posts/python-performance-profiling/) has some guides on doing so with cProfile and snakeviz

**Comment by danbodily** [comment]
Slow Mini Conda 3.8.10
![Slow(Miniconda3 8 10)](https://user-images.githubusercontent.com/9097953/235743730-f342ae95-403c-45db-ac9a-a76a2fa1a9e3.png)
Faster Python 3.8.10 from Pipy
![Fast(PiPyPython3 8 10)](https://user-images.githubusercontent.com/9097953/235743741-32aff3b4-6477-41da-97e6-cf7fa509a6a6.png)

Good suggestion - I've included the trees here - For whatever reason Miniconda's 3.8.10 isn't using pyembree maybe?  These trees are always difficult to understand.

**Comment by danbodily** [comment]
Indeed in stepping in pdb, it appears that the fast python version that comes native has makes:

tmesh.ray.intersects_location  bound to
 <bound method RayMeshIntersector.intersects_location of <trimesh.ray.ray_pyembree.RayMeshIntersector object at 0x00000207C270B850>>

Whereas in the other slower conda installed version  has bound to the trimesh intersector: 
<bound method RayMeshIntersector.intersects_location of <trimesh.ray.ray_triangle.RayMeshIntersector object at 0x000001F588725CA0>

This is in \pyvista\\core\\filters\\poly_data.py line 2159.

Digging in deeper I found this failed in Miniconda and succeeded in the normal install:

ImportError: DLL load failed while importing rtcore_scene: The specified module could not be found.

This is related to https://github.com/pyvista/pyvista/issues/1529  and https://github.com/mikedh/trimesh/issues/875

I pip installed pyembree and am still having issues using Miniconda's 3.8.10 distribution 

Ultimately my solution was to ditch Miniconda and go to native python using venv,  everything checks out fine.
